pkg/agent/sysadvisor/plugin/qosaware/resource/memory/headroompolicy/policy_numa_aware.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/kubewharf/katalyst-core/pkg/util/machine"
+	"github.com/kubewharf/katalyst-core/pkg/util/metric"
+)
+
+type PolicyNUMAAware struct {
+	*PolicyBase
+
+	// memoryHeadroom is valid to be used iff updateStatus successes
+	memoryHeadroom float64
+	updateStatus   types.PolicyUpdateStatus
+
+	conf *config.Configuration
+}
+
+func NewPolicyNUMAAware(conf *config.Configuration, _ interface{}, metaReader metacache.MetaReader,
+	metaServer *metaserver.MetaServer, _ metrics.MetricEmitter) HeadroomPolicy {
+	p := PolicyNUMAAware{
+		PolicyBase:   NewPolicyBase(metaReader, metaServer),
+		updateStatus: types.PolicyUpdateFailed,
+		conf:         conf,
+	}
+
+	return &p
+}
+
+func (p *PolicyNUMAAware) Name() types.MemoryHeadroomPolicyName {
+	return types.MemoryHeadroomPolicyNUMAAware
+}
+
+func (p *PolicyNUMAAware) reclaimedContainersFilter(ci *types.ContainerInfo) bool {
+	return ci != nil && ci.QoSLevel == apiconsts.PodAnnotationQoSLevelReclaimedCores
+}
+
+func (p *PolicyNUMAAware) Update() (err error) {
+	defer func() {
+		if err != nil {
+			p.updateStatus = types.PolicyUpdateFailed
+		} else {
+			p.updateStatus = types.PolicyUpdateSucceeded
+		}
+	}()
+
+	var (
+		errList           []error
+		reclaimableMemory float64
+		data              metric.MetricData
+	)
+
+	availNUMAs := p.metaServer.CPUDetails.NUMANodes()
+
+	reclaimedCoresContainers := make([]*types.ContainerInfo, 0)
+	p.metaReader.RangeContainer(func(podUID string, containerName string, containerInfo *types.ContainerInfo) bool {
+		if p.reclaimedContainersFilter(containerInfo) {
+			reclaimedCoresContainers = append(reclaimedCoresContainers, containerInfo)
+			return true
+		}
+
+		nodeReclaim := p.conf.GetDynamicConfiguration().EnableReclaim
+		reclaimEnable, err := helper.PodEnableReclaim(context.Background(), p.metaServer, podUID, nodeReclaim)
+		if err != nil {
+			errList = append(errList, err)
+			return true
+		}
+
+		if containerInfo.IsNumaExclusive() && !reclaimEnable {
+			memset := machine.GetCPUAssignmentNUMAs(containerInfo.TopologyAwareAssignments)
+			if memset.IsEmpty() {
+				errList = append(errList, fmt.Errorf("container(%v/%v) TopologyAwareAssignments is empty", containerInfo.PodName, containerName))
+				return true
+			}
+			availNUMAs = availNUMAs.Difference(memset)
+		}
+		return true
+	})
+
+	err = errors.NewAggregate(errList)
+	if err != nil {
+		return err
+	}
+
+	for _, numaID := range availNUMAs.ToSliceInt() {
+		data, err = p.metaServer.GetNumaMetric(numaID, consts.MetricMemFreeNuma)
+		if err != nil {
+			return err
+		}
+		numaFree := data.Value
+
+		reclaimedCoresUsed := 0.
+		for _, container := range reclaimedCoresContainers {
+			data, err = p.metaServer.GetContainerNumaMetric(container.PodUID, container.ContainerName, strconv.Itoa(numaID), consts.MetricsMemTotalPerNumaContainer)
+			if err != nil {
+				return err
+			}
+			reclaimedCoresUsed += data.Value
+		}
+
+		reclaimableMemory += numaFree + reclaimedCoresUsed
```

Comment:
Is it safer to substract shared/dedicated cores usage buffer, i.e. usage * 10%?

Reply:
Maybe reclaimed_cores potential usage(the gap between memory limit and actual usage) should also be considered.